File: python/src/azex/types/messages.py
"""Types for Anthropic-compatible messages API."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional, Union

from pydantic import BaseModel


class TextBlock(BaseModel):
    type: Literal["text"]
    text: str


class ThinkingBlock(BaseModel):
    type: Literal["thinking"]
    thinking: str
    signature: Optional[str] = None


class ToolUseBlock(BaseModel):
    type: Literal["tool_use"]
    id: str
    name: str
    input: Dict[str, Any]


ContentBlock = Union[TextBlock, ThinkingBlock, ToolUseBlock]


class MessageUsage(BaseModel):
    input_tokens: int
    output_tokens: int


class Message(BaseModel):
    id: str
    type: Literal["message"]
    role: Literal["assistant"]
    content: List[ContentBlock]
    model: str
    stop_reason: Optional[str] = None
    stop_sequence: Optional[str] = None
    usage: MessageUsage


# --- Stream events ---

class TextDelta(BaseModel):
    type: Literal["text_delta"]
    text: str


class ThinkingDelta(BaseModel):
    type: Literal["thinking_delta"]
    thinking: str


class InputJSONDelta(BaseModel):
    type: Literal["input_json_delta"]
    partial_json: str


Delta = Union[TextDelta, ThinkingDelta, InputJSONDelta]


class MessageStartEvent(BaseModel):
    type: Literal["message_start"]
    message: Message


class ContentBlockStartEvent(BaseModel):
    type: Literal["content_block_start"]
    index: int
    content_block: ContentBlock


class ContentBlockDeltaEvent(BaseModel):
    type: Literal["content_block_delta"]
    index: int
    delta: Delta


class ContentBlockStopEvent(BaseModel):
    type: Literal["content_block_stop"]
    index: int


class MessageDeltaEvent(BaseModel):
    type: Literal["message_delta"]
    delta: Dict[str, Any]
    usage: Optional[MessageUsage] = None


class MessageStopEvent(BaseModel):
    type: Literal["message_stop"]


MessageStreamEvent = Union[
    MessageStartEvent,
    ContentBlockStartEvent,
    ContentBlockDeltaEvent,
    ContentBlockStopEvent,
    MessageDeltaEvent,
    MessageStopEvent,
]

File: typescript/src/types/messages.ts
export interface TextBlock {
  type: "text";
  text: string;
}

export interface ThinkingBlock {
  type: "thinking";
  thinking: string;
  signature?: string;
}

export interface ToolUseBlock {
  type: "tool_use";
  id: string;
  name: string;
  input: Record<string, unknown>;
}

export type ContentBlock = TextBlock | ThinkingBlock | ToolUseBlock;

export interface MessageUsage {
  input_tokens: number;
  output_tokens: number;
}

export interface Message {
  id: string;
  type: "message";
  role: "assistant";
  content: ContentBlock[];
  model: string;
  stop_reason: string | null;
  stop_sequence: string | null;
  usage: MessageUsage;
}

export interface TextDelta {
  type: "text_delta";
  text: string;
}

export interface ThinkingDelta {
  type: "thinking_delta";
  thinking: string;
}

export interface InputJSONDelta {
  type: "input_json_delta";
  partial_json: string;
}

export type Delta = TextDelta | ThinkingDelta | InputJSONDelta;

export interface MessageStartEvent {
  type: "message_start";
  message: Message;
}

export interface ContentBlockStartEvent {
  type: "content_block_start";
  index: number;
  content_block: ContentBlock;
}

export interface ContentBlockDeltaEvent {
  type: "content_block_delta";
  index: number;
  delta: Delta;
}

export interface ContentBlockStopEvent {
  type: "content_block_stop";
  index: number;
}

export interface MessageDeltaEvent {
  type: "message_delta";
  delta: Record<string, unknown>;
  usage?: MessageUsage;
}

export interface MessageStopEvent {
  type: "message_stop";
}

export type MessageStreamEvent =
  | MessageStartEvent
  | ContentBlockStartEvent
  | ContentBlockDeltaEvent
  | ContentBlockStopEvent
  | MessageDeltaEvent
  | MessageStopEvent;

export interface MessageCreateParams {
  model: string;
  messages: Array<{ role: "user" | "assistant"; content: string | ContentBlock[] }>;
  max_tokens: number;
  system?: string | Array<{ type: string; text: string }>;
  temperature?: number;
  top_p?: number;
  top_k?: number;
  stop_sequences?: string[];
  tools?: unknown[];
  tool_choice?: unknown;
  stream?: boolean;
}
